File: HW/HW12/HW12.java
package HW12;

import java.util.*;

public class HW12 {
    public int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {
        // Create a list to store all pipes including wells
        List<int[]> allPipes = new ArrayList<>();
        
        // Add all wells to the list as pipes with cost wells[i]
        for (int i = 0; i < n; i++) {
            allPipes.add(new int[]{0, i + 1, wells[i]});
        }
        
        // Add all given pipes to the list
        for (int[] pipe : pipes) {
            allPipes.add(pipe);
        }
        
        // Sort all pipes based on their cost
        Collections.sort(allPipes, (a, b) -> a[2] - b[2]);
        
        // Initialize variables to track minimum cost and union-find structure
        int minCost = 0;
        UnionFind uf = new UnionFind(n + 1);
        
        // Iterate through all pipes
        for (int[] pipe : allPipes) {
            int house1 = pipe[0];
            int house2 = pipe[1];
            int cost = pipe[2];
            
            // If the two houses are not already connected, connect them and add cost
            if (!uf.connected(house1, house2)) {
                uf.union(house1, house2);
                minCost += cost;
            }
        }
        
        return minCost;
    }

    public static void main(String[] args) {
        HW12 solution = new HW12();

        // Test Case 1
        int n1 = 2;
        int[] wells1 = {1, 1};
        int[][] pipes1 = {{1, 2, 1}, {1, 2, 2}};
        int result1 = solution.minCostToSupplyWater(n1, wells1, pipes1);
        System.out.println("Test Case 1: "  + result1);
    }

    // Union-Find (Disjoint Set) data structure implementation
    class UnionFind {
        int[] parent;

        public UnionFind(int n) {
            parent = new int[n];
            for (int i = 0; i < n; i++) {
                parent[i] = i;
            }
        }

        public int find(int x) {
            if (parent[x] != x) {
                parent[x] = find(parent[x]);
            }
            return parent[x];
        }

        public boolean connected(int x, int y) {
            return find(x) == find(y);
        }

        public void union(int x, int y) {
            int rootX = find(x);
            int rootY = find(y);
            parent[rootX] = rootY;
        }
    }
}

File: HW/HW12/HW12.py
from typing import List

class Solution:
    def minCostToSupplyWater(self, n: int, wells: List[int], pipes: List[List[int]]) -> int:
        #list of all pipes and wells
        all_pipes = [[0, i + 1, wells[i]] for i in range(n)]
        
        # Add all pipes to list
        all_pipes.extend(pipes)
        
        # Sort pipes based on cost
        all_pipes.sort(key=lambda x: x[2])
        
        # Initialize variables to track minimum cost and union-find structure
        min_cost = 0
        parent = list(range(n + 1))

        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]

        def union(x, y):
            root_x = find(x)
            root_y = find(y)
            parent[root_x] = root_y

        # Iterate through all pipes
        for pipe in all_pipes:
            house1, house2, cost = pipe
            
            # If the two houses are not already connected, connect them and add cost
            if find(house1) != find(house2):
                union(house1, house2)
                min_cost += cost

        return min_cost

# Test Cases
solution = Solution()

# Test Case
n = 2
wells = [1, 1]
pipes = [[1, 2, 1], [1, 2, 2]]
result = solution.minCostToSupplyWater(n, wells, pipes)
print("Test Case:", result)
